server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java

```diff
@@ -62,6 +72,46 @@ protected void unpromotableShardOperation(
         Task task,
         UnpromotableShardRefreshRequest request,
         ActionListener<ActionResponse.Empty> responseListener
+    ) {
+        if (clusterService.state().version() >= request.getClusterStateVersion()) {
+            doUnpromotableShardOperation(request, responseListener);
+        } else {
+            final Thread initialThread = Thread.currentThread();
+            ClusterStateObserver.waitForState(
+                clusterService,
+                clusterService.threadPool().getThreadContext(),
+                new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        if (initialThread == Thread.currentThread()) {
+                            doUnpromotableShardOperation(request, responseListener);
```

Comment:
To fix the issue observed in test, we really just need the version barrier when either indexService or indexShard is `null`. But I feel applying the barrier consistently is better. But I am happy to be convinced otherwise.

Reply:
Would it be an option to just allow the shard to not exist and respond "OK", even without a version barrier?

Reply:
I thought about this. But I was not sure whether it is safe in all cases. But maybe it is. The issue happens when the IndexShard is not available, i.e. before `IndexShard#createShard` is completed. The `indexShard#preRecovery` happens strictly after it and should see any refresh changes when it contacts the indexing shard. So it seems we are OK to just respond OK without anymore additional logic. Feels a bit weird. But should be safe?

Reply:
If we just respond OK, from end user's perspective, a refresh response can say 2 shards are successfully refreshed. But 1 of them could be a no-op due to shard still being created. I guess that's OK since it is momentary?

Reply:
It is still only going to respond with refreshed data. How we ensure that is not relevant to the user I think. The end result is the same.

Reply:
I think what you describe would be a problem with the registration process. Once the shard is indeed successfully relocated, we should stop accepting registrations. That is before the target shard is started. Allowing registrations to proceed on an old primary could lead to serving unfresh data regardless of the observer here. I am unsure if we have any such protection (did not check).

> Thus, the search node is finally on an older commit than what the refresh request wanted. And may serve outdated data

We should not accept any new data on the target until the target is started (thus this is only a problem after the shard is started, the administrative commit done during initialization is not important to data serving).

Reply:
Makes sense to me to do the fixes in that order.

Reply:
Thanks for the discussion. I learnt from it. As suggested, I raised serverless#2277 to reject commit registration once the indexing shard is marked as relocated with a test case that demonstrates currently we do allow registration after relocated.